### PR TITLE
Introduce more loops when inlining

### PR DESCRIPTION
**Pull request by MatthewFluet**
When inlining potentially recursive functions, we would like to convert a tail call of a previously inlined function into a goto to the beginning of the inlined code.  However, this tail call is nested inside the non-tail call of a different function, then the tail call cannot be converted to a goto.

Previously, we tracked only one `return: Return.t` that corresponded to the "return context" through all of the function contexts (the "root" function and all of the nested inlined functions).  If this were `Tail`, then a new tail call to a function in the context could be converted to a goto.  However, if this were `NonTail`, then we unconditionally retained the call (with the composed return context).  This misses some opportunities, because if the `NonTail`-ness is "above" the occurrence of this function in the context, then the call *can* be converted to a goto.

Now, we track one `Return.t` per function in the context.  If a call is tail and all of the `Return.t`s up to (but not including) a match of the called function in the function context are `Tail`, then it is converted to a goto.

See MPLLang/mpl#203

### CODE REVIEW

**Comment by colin-mcd** [comment]
Thanks Matthew! This looks great

**Comment by colin-mcd** [comment]
Wait nvm, this change causes the error `Fail: Vector.foldi2From` (from lib/mlton/basic/vector.fun)... Any ideas how?

**Comment by MatthewFluet** [comment]
I suspect that I got the folding of the composed return wrong; I'm looking into it.

**Comment by MatthewFluet** [comment]
Yes, I had the composition  backwards.  Seems to be working now.
(The `Fail: Vector.foldi2From` probably came from something in the `Shrink` pass, where the mis-composed return meant that there was a mismatch in expected vs actual args somewhere.)

File: mlton/ssa/inline.fun [hunk 1]
```diff
@@ -474,17 +486,28 @@ fun transform {program as Program.T {datatypes, globals, functions, main},
...
-                                 (case return of
-                                     Return.Tail => new (Transfer.Goto
-                                                         {dst = start,
-                                                          args = args})
-                                   | _ => doCall ())
+                           case findLoop (nest, Return.isTail return') of
+                              SOME (SOME start) => new (Transfer.Goto
+                                                        {dst = start,
+                                                         args = args})
+                            | SOME NONE => doCall ()
```
[the remark this answers is not in the document]
Comment (MatthewFluet) [reply mlton/ssa/inline.fun:510]: Sure, that's reasonable, though, as you noted previously, can lead to infinite regress.

**Comment by MatthewFluet** [comment]
Updated to obey a `__inline_always__` annotation even if the function has been previously inlined.

This particular logic makes an `__inline_never__` annotation on a call site "win" over an "__inline_always__` annotation on the function, but I could be convinced that `always` should win over `never`.

**Comment by colin-mcd** [comment]
Just pushed some changes that effectively implement this, among other things: 7e0e055

**Comment by MatthewFluet** [comment]
Forcing polyvariance duplication with `__inline_always__` is reasonable; but, I do think that there are some other benefits to what was being done here, with the conversion of a `Tail` call in a `Tail` context to a previously inlined function into a `Goto`.  I'll see about rebasing the changes.

**Comment by MatthewFluet** [comment]
O.k.  I looked more carefully at 7e0e055 and the diff with this branch and see that it is mostly equivalent.  There seems to be an unnecessary argument to `fun findLoops` in `inline.fun`.

(BTW, I think the git history would be a little cleaner if these changes to `inline.fun` were committed with this commit message, rather than combined with the polyvariance changes (and a commit message that only mentions the polyvariance).)

**Comment by MatthewFluet** [comment]
Closing.

**Pull request closed without merging**